common/src/main/java/com/wynntils/core/features/overlays/OverlayManager.java

```diff
@@ -45,35 +55,143 @@ public OverlayManager(CrashReportManager crashReportManager) {
+        overlayParentMap.getOrDefault(parent, Collections.emptyList()).forEach(this::enableOverlay);
+    }
+
+    public void enableOverlay(Overlay enableOverlay) {
+        if (!enableOverlay.isEnabled()) return;
 
-        this.enabledOverlays.addAll(enabledOverlays);
+        this.enabledOverlays.add(enableOverlay);
 
         enabledOverlays.forEach(
                 overlay -> overlay.getConfigOptionFromString("userEnabled").ifPresent(overlay::onConfigUpdate));
     }
 
+    public void discoverOverlays(Feature feature) {
+        Field[] overlayFields = FieldUtils.getFieldsWithAnnotation(feature.getClass(), OverlayInfo.class);
+        for (Field overlayField : overlayFields) {
+            try {
+                Object fieldValue = FieldUtils.readField(overlayField, feature, true);
+
+                if (!(fieldValue instanceof Overlay overlay)) {
+                    throw new RuntimeException("A non-Overlay class was marked with OverlayInfo annotation.");
+                }
+
+                OverlayInfo annotation = overlayField.getAnnotation(OverlayInfo.class);
+                Managers.Overlay.registerOverlay(
+                        overlay, feature, annotation.renderType(), annotation.renderAt(), annotation.enabled());
+
+                assert !overlay.getTranslatedName().startsWith("feature.wynntils.");
+            } catch (IllegalAccessException e) {
+                WynntilsMod.error("Unable to get field " + overlayField, e);
+            }
+        }
+    }
+
+    public void discoverOverlayGroups(Feature feature) {
+        List<OverlayGroupHolder> holders = Stream.of(feature.getClass().getDeclaredFields())
+                .filter(f -> f.isAnnotationPresent(OverlayGroup.class))
+                .map(field -> {
+                    OverlayGroup annotation = field.getAnnotation(OverlayGroup.class);
+                    return new OverlayGroupHolder(
+                            field, feature, annotation.renderType(), annotation.renderAt(), annotation.instances());
+                })
+                .toList();
+
+        holders.forEach(this::createOverlayGroupWithDefaults);
+
+        overlayGroupMap.put(feature, holders);
+    }
+
+    // endregion
+
+    // region Overlay Groups
+
+    public void createOverlayGroupWithDefaults(OverlayGroupHolder holder) {
+        recreateGroupOverlaysWithIds(
+                holder,
+                IntStream.rangeClosed(1, holder.getDefaultCount()).boxed().toList());
+    }
+
+    public void createOverlayGroupWithIds(OverlayGroupHolder holder, List<Integer> ids) {
+        recreateGroupOverlaysWithIds(holder, ids);
+    }
+
+    public int extendOverlayGroup(OverlayGroupHolder holder) {
+        List<Integer> ids = holder.getOverlays().stream()
+                .map(overlay -> ((DynamicOverlay) overlay).getId())
+                .collect(Collectors.toList());
+        ;
+
+        int newId = 1;
+        while (ids.contains(newId)) {
```

Comment:
My CS knee-jerk reaction tells me "it's growing quadratically!" but I guess it's fine in this case...